recipe/linear_algebra/util.cc

```diff
@@ -0,0 +1,31 @@
+#include <algorithm>
+#include <initializer_list>
+#include <memory>
+
+namespace recipe {
+namespace linear_algebra {
+
+std::unique_ptr<double[]> MakeDoubleArray(std::initializer_list<double> list) {
```

Comment:
I may not understand your questions precisely. I assume you are asking me the reasons why I use `std::initializer_list` to create `double[]`. The reason is as follows;

(1) `double[]` and its unique_ptr are widely used as function arguments in the codebase.
To test such functions, we need to generate the types.

```cpp
std::unique_ptr<double[]> data = MakeDoubleArray({
              1, 1, 1,
              0, 1, 1,
              0, 0, 1,
              });
```
is much more readable than

```cpp
std::unique_ptr<double[]> data(new double[9]);
data[0] = 1;
data[1] = 1;
data[2] = 1;
data[3] = 0;
data[4] = 1;
data[5] = 1;
data[6] = 0;
data[7] = 0;
data[8] = 1;
```

Also, the latter implementation cannot declare the variable `data` with `const` qualifier.

If you mean `std::initializer_list` is not an appropriate type for the arguments, in that case, the reasons are two folds;

(2)

```cpp
std::unique_ptr<double[]> data1({1, 1, 1});  // error
double[] data2{1, 1, 1};  // ok
double[] data3 = {1, 1, 1};  // ok
std::unique_ptr<double[]> data22(data2);  // malloc error: data2 is a static array so it cannot be deleted
std::unique_ptr<double[]> data32(data3);  // malloc error: the same reason above
```

We can use a custom deleter for `data22` and `data32` above to avoid malloc error, but it would make the code more complicated.

(3)

```cpp
std::unique_ptr<double[]> MakeDoubleArray(std::vector<double> data) {
// implement
}
std::unique_ptr<double[]> data = MakeDoubleArray({1, 2, 3});
// OK but the funciton invokes std::vector::vector(std::initializer_list<double>) constructor
// so the unnecessary conversion std::initializer_list<double> -> std::vector<double> occurs
```

I'm not sure whether the conversion from `std::vector<double>` to `std::unique_ptr<double[]>` will be necessary or not.
Possible use cases of `std::vector<double>` are

* Reading or importing data from fiels or inputs as `std::vector<T>`
* Converting `std::vector<T>` to `Matrix` or `Vector` class.

If `std::vector<T>` need to be converted to `std::unique_ptr<T>` with some specific use cases, we can implement it then.